File: validateBinarySearch.py
# Definition for a binary tree node.
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    # def isValidBST(self, root: TreeNode) -> bool:
    #     stack =[]
    #     prev=None
    #     while root or stack:
    #         while root is not None:
    #             stack.append(root)
    #             root=root.left
    #             print(root)
    #         print(root)
    #
    #         root=stack.pop()
    #
    #         if prev and prev.val>root.val:
    #             return False
    #         prev = root
    #
    #         print(root.val)
    #         root=root.right
    #     return True
    def isvalidrecBST(self,root: TreeNode) -> None:
        isValid=True
        isValid=self.isValidBST(root=root)
        return isValid


    ptr=None
    def isValidBST(self, root: TreeNode) -> bool:
        if root is None:
            return True
        self.isValidBST(root.left)
        print(root.val)
        if self.ptr is not None and self.ptr.val > root.val:
            return False
        self.ptr=root
        return  self.isValidBST(root.right)





if __name__ == '__main__':

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.right.left = TreeNode(5)
    root.right.right = TreeNode(6)
    root.right.left.left = TreeNode(7)
    root.right.left.right = TreeNode(8)

    print(Solution().isValidBST(root=root))

File: ValidateBinarySearch.java
// Time Complexity : O(N), N is the max number of elements in the tree
// Space Complexity : O(H) where H is the height of the tree. This will be the max amount of space used in the stack.
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode prev;
    boolean isValid;
    public boolean isValidBST(TreeNode root) {
        // base case
        if (root == null) return isValid;

        isValid = true;
        inOrderTraversal(root);

        return isValid;
    }

    private void inOrderTraversal(TreeNode root) {
        // base case
        if (root == null) return;

        // Left sub-tree
        if (isValid) inOrderTraversal(root.left);
        if (prev !=null && prev.val >= root.val) {
            isValid = false;
            return;
        }
        prev = root;
        // Right sub-tree
        if (isValid) inOrderTraversal(root.right);
    }
}